include/CL/sycl/parallelism/detail/parallelism.hpp

```diff
@@ -239,35 +239,47 @@ void parallel_for_workitem(const group<Dimensions> &g,
 
      The issue is that the parallel_for_workitem() execution is slow even
      when nd_item::barrier() is not used
-
-     \todo Simplify by just using omp parallel for collapse
   */
-
+  T_Item index{g.get_nd_range()};
   range<Dimensions> l_r = g.get_nd_range().get_local_range();
-  auto tot = l_r.get(0);
-  for (int i = 1; i < (int) Dimensions; ++i) {
-    tot *= l_r.get(i);
-  }
-#pragma omp parallel num_threads(tot)
-  {
-    T_Item index { g.get_nd_range() };
-    id<Dimensions> local; // to initialize correctly
-#pragma omp for nowait
-    for (std::size_t th_id = 0; th_id < tot; ++th_id) {
-      if (Dimensions == 1) {
-        local[0] = th_id;
-      } else if (Dimensions == 2) {
-        local[0] = th_id / l_r.get(1);
-        local[1] = th_id % l_r.get(1);
-      } else if (Dimensions == 3) {
-        local[0] = th_id / (l_r.get(1)*l_r.get(2));
-        local[1] = (th_id / l_r.get(2)) % l_r.get(1);
-        local[2] = th_id % l_r.get(2);
-      }
-      index.set_local(local);
-      index.set_global(local + id<Dimensions>(l_r)*g.get_id());
+  id<Dimensions> temp, id_l_r = id<Dimensions>(l_r);
+
+  if constexpr (Dimensions == 1) {
+#pragma omp parallel for collapse(1) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i) {
+      temp[0] = i;
+      index.set_local(temp);
+      temp[0] = i + id_l_r[0] * g.get_id()[0];
+      index.set_global(temp);
       f(index);
     }
+  } else if constexpr (Dimensions == 2) {
+#pragma omp parallel for collapse(1) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i) {
+      #pragma omp simd collapse(1)
+      for (size_t j = 0; j < l_r.get(1); ++j) {
+        temp[0] = i; temp[1] = j;
+        index.set_local(temp);
+        temp[0] = i + id_l_r[0] * g.get_id()[0];
+        temp[1] = j + id_l_r[1] * g.get_id()[1];
+        index.set_global(temp);
+        f(index);
+      }
+    }
+  } else if constexpr (Dimensions == 3) {
+    #pragma omp parallel for collapse(2) firstprivate(index, temp)
+    for (size_t i = 0; i < l_r.get(0); ++i)
+      for (size_t j = 0; j < l_r.get(1); ++j)
+      #pragma omp simd collapse(1)
+        for (size_t k = 0; k < l_r.get(2); ++k) {
+          temp[0] = i; temp[1] = j; temp[2] = k;
+          index.set_local(temp);
+          temp[0] = i + id_l_r[0] * g.get_id()[0];
+          temp[1] = j + id_l_r[1] * g.get_id()[1];
+          temp[2] = k + id_l_r[2] * g.get_id()[2];
+          index.set_global(temp);
```

Comment:
```
 index.set_global({ i + id_l_r[0] * g.get_id()[0], 
                    j + id_l_r[1] * g.get_id()[1],
                    k + id_l_r[2] * g.get_id()[2] });
```
or
```
index.set_global(get_local_id() + id_l_r*g.get_id());
```
?

Reply:
Have you tried with `index.set_global(get_local_id() + id_l_r*g.get_id());` by the way ?

Reply:
Tried it and it works (and its the current implementation now), I originally went with the other as I thought it might be a bit faster as it could be simpler for the OpenMP compiler to use for SIMD calculations.. it apparently isn't (or at least not noticeably so)! I should have more faith in modern compiler optimizations it seems.